snuba/datasets/discover.py

```diff
@@ -0,0 +1,247 @@
+from abc import ABC
+from datetime import timedelta
+from typing import Mapping, Sequence
+
+from snuba.clickhouse.columns import (
+    Array,
+    ColumnSet,
+    DateTime,
+    FixedString,
+    Nested,
+    Nullable,
+    String,
+    UInt,
+    UUID,
+)
+from snuba.datasets.dataset import TimeSeriesDataset
+from snuba.datasets.dataset_schemas import DatasetSchemas
+from snuba.datasets.factory import get_dataset
+from snuba.datasets.schemas import Schema
+from snuba.query.extensions import QueryExtension
+from snuba.query.parsing import ParsingContext
+from snuba.query.project_extension import ProjectExtension, ProjectWithGroupsProcessor
+from snuba.query.query import Query
+from snuba.query.query_processor import QueryProcessor
+from snuba.query.timeseries import TimeSeriesExtension
+from snuba.request.request_settings import RequestSettings
+from snuba.util import is_condition
+
+EVENTS = 'events'
+TRANSACTIONS = 'transactions'
+
+
+def detect_dataset(query: Query) -> str:
+    """
+    Given a query, we attempt to guess whether it is better to fetch data from the
+    "events" or "transactions" dataset. This is going to be wrong in some cases.
+    """
+    # First check for a top level condition that matches either type = transaction
+    # type != transaction.
+    conditions = query.get_conditions()
+    if conditions:
+        for idx, condition in enumerate(conditions):
+            if is_condition(condition):
+                if tuple(condition) == ('type', '!=', 'transaction'):
+                    return EVENTS
+                elif tuple(condition) == ('type', '=', 'transaction'):
+                    return TRANSACTIONS
```

Comment:
conditions can be nested:
```
[
    [
        ['w', '=', '1'], ['x', '=', '2']
    ],
    [
        ['y', '=', '3'], ['z', '=', '4']
    ],
]

(w = '1' OR x = '2') AND (y = '3' OR z = '4')
```

Reply:
This is a very limited dataset detection function, which matches the logic that is in Sentry right now. If there is a nested OR condition on type, we still cannot be sure of the dataset so we do not handle that case in this way.

Reply:
fair point.